src/tools/interop/idt/capture/platform/android/android.py

```diff
@@ -0,0 +1,220 @@
+        self.logcat_proc: BashRunner | None = None
+        self.logcat_command = f'adb -s {self.device_id} logcat -T 1 >> {self.logcat_output_path}'
+
+        # TODO: BT Logs
+
+        screen_cast_name = create_standard_log_name('screencast', 'mp4')
+        self.screen_cap_output_path = os.path.join(
+            self.artifact_dir, screen_cast_name)
+        self.screen_proc: BashRunner | None = None
+        self.check_screen_command = "shell dumpsys deviceidle | grep mScreenOn"
+        self.screen_path = f'/sdcard/Movies/{screen_cast_name}'
+        self.screen_command = f'adb -s {self.device_id} shell screenrecord {self.screen_path}'
+        self.screen_recording_already_pulled = False
+        self.screen_pull_command = f'pull {self.screen_path} {self.screen_cap_output_path}'
+
+    def run_adb_command(
+            self,
+            command: str,
+            capture_output: bool = False) -> BashRunner:
+        """
+        Run an adb command synchronously
+        Capture_output must be true to call get_captured_output() later
+        """
+        return BashRunner(
+            f'adb -s {self.device_id} {command}',
+            sync=True,
+            capture_output=capture_output)
+
+    def get_adb_devices(self) -> typing.Dict[str, bool]:
+        """Returns a dict of device ids and whether they are authorized"""
+        adb_devices = BashRunner('adb devices', sync=True, capture_output=True)
+        adb_devices_output = adb_devices.get_captured_output().split('\n')
+        devices_auth = {}
+        header_done = False
+        for line in adb_devices_output:
+            if header_done:
+                line_parsed = line.split("\t")
+                device_id = line_parsed[0]
+                device_is_auth = line_parsed[1] == "device"
+                if line_parsed[1] == "offline":
+                    disconnect_command = f"adb disconnect {device_id}"
+                    print(f"Device {device_id} is offline, trying disconnect!")
+                    BashRunner(
+                        disconnect_command,
+                        sync=True,
+                        capture_output=False)
+                else:
+                    devices_auth[device_id] = device_is_auth
+            header_done = True
+        self.adb_devices = devices_auth
+        return self.adb_devices
+
+    def _only_one_device_connected(self) -> bool:
+        return len(self.adb_devices) == 1
+
+    def _get_first_connected_device(self) -> str:
+        return list(self.adb_devices.keys())[0]
+
+    def _set_device_if_only_one_connected(self) -> None:
+        if self._only_one_device_connected():
+            self.device_id = self._get_first_connected_device()
+            print(f'Only one device detected; using {self.device_id}')
+
+    def _log_adb_devices(self) -> None:
+        for dev in self.adb_devices:
+            print(dev)
+
+    @staticmethod
+    def _is_connection_str(adb_input_str: str) -> bool:
+        valid_ipv4 = False
```

Comment:
code looks fine, but is this guaranteed to be ipv4?

Reply:
I have personally only ever seen IPv4 listed in the "IP address & Port" section of Android Wireless debugging.